tests/test_simulator_utils.py

```diff
@@ -0,0 +1,226 @@
+from typing import Callable
+import pytest
+import torch
+from torch.distributions import Uniform, MultivariateNormal, Distribution
+from sbi.simulators.simutils import (
+    prepare_sbi_problem,
+    process_prior,
+    process_observed_data,
+    process_simulator,
+    ScipyToPytorchWrapper,
+    CustomToPytorchWrapper,
+)
+from scipy.stats import multivariate_normal, uniform, beta
+from sbi.simulators.linear_gaussian import linear_gaussian
+from sbi.utils.torchutils import BoxUniform
+import numpy as np
+
+
+class UserUniform:
+    """User defined numpy uniform prior. 
+    
+    Used for testing to mimick a user-defined prior with valid .sample and .log_prob 
+    methods. 
+    """
+
+    def __init__(self, lower, upper, return_numpy=False):
+        self.lower = lower
+        self.upper = upper
+        self.dist = BoxUniform(lower, upper)
+        self.return_numpy = return_numpy
+
+    def sample(self, sample_shape=torch.Size([])):
+        samples = self.dist.sample(sample_shape)
+        return samples.numpy() if self.return_numpy else samples
+
+    def log_prob(self, values):
+        if self.return_numpy:
+            values = torch.as_tensor(values)
+        log_probs = self.dist.log_prob(values)
+        return log_probs.numpy() if self.return_numpy else log_probs
+
+
+def torch_simulator_no_batch(x):
+    """Identity simulator throwing assertion error when called on a batch."""
+    assert x.ndim == 1, "cant handle batches, bitches."
+    return x
+
+
+def numpy_simulator(x):
+    """Simulator returning zeroed input, Assertion error if input is not numpy."""
+    assert isinstance(x, np.ndarray)
+    return np.zeros_like(x)
+
+
+def numpy_linear_gaussian(theta):
+    """Linear Gaussian simulator wrapped to get and return numpy."""
+    return linear_gaussian(torch.as_tensor(theta, dtype=torch.float32)).numpy()
+
+
+def list_simulator(x):
+    return list(x)
+
+
+@pytest.mark.parametrize(
+    "wrapper, prior",
+    (
+        (
+            CustomToPytorchWrapper,
+            UserUniform(torch.zeros(3), torch.ones(3), return_numpy=True),
+        ),
+        (ScipyToPytorchWrapper, multivariate_normal()),
+        (ScipyToPytorchWrapper, uniform()),
+        (ScipyToPytorchWrapper, beta(a=1, b=1)),
+    ),
+)
+def test_prior_wrappers(wrapper, prior):
+    """Test prior wrappers to pytorch distributions."""
+    prior = wrapper(prior)
+
+    theta = prior.sample((10,))
+    assert isinstance(theta, torch.Tensor)
+    assert theta.shape[0] == 10
+
+    log_probs = prior.log_prob(theta)
```

Comment:
Maybe add a little comment as to how theta has now `batch_size=10` and is beng scored all at once? thinking abou reading this in 6 months...